### PR TITLE
[BugFix] fix the inconsistent behavior of JDBC external table with version 2.4

### LINKED ISSUE #14235: jdbc extertnal table query result is error
### Steps to reproduce the behavior (Required)
1. char and varchar type , length is different
8627a608-63ed-4088-b5ef-f2145e0cf664
mysql 
image

2. oracle float
image

3. pg timestamp
image

### Expected behavior (Required)
1. error
2. sucess
3. without ms
### Real behavior (Required)
1.mysql> select * from jdbc_ext_tbl_mysql_type_char_test;
+-----------+
| type_char |
+-----------+
| abcdefgh  |
+-----------+
1 row in set (0.01 sec)

mysql> select * from jdbc_ext_tbl_mysql_type_varchar_test;
+--------------+
| type_varchar |
+--------------+
| abcdefgh     |
+--------------+
1 row in set (0.02 sec)

2. failed

3.
mysql> select * from jdbc_ext_tbl_pg_type_timestamp_test;
+----------------------------+----------------------------+
| type_timestamp             | type_timestamp2            |
+----------------------------+----------------------------+
| 0001-01-01 00:00:00.789000 | 0001-12-31 23:54:17.798000 |
| 9999-12-31 10:00:00.001000 | 9999-12-31 10:00:00.001000 |
+----------------------------+----------------------------+
2 rows in set (0.02 sec)
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
mysql> select current_version();
+----------------------------+
| current_version()          |
+----------------------------+
| BRANCH-2.5-RELEASE 1e7e914 |
+----------------------------+
1 row in set (0.01 sec)

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #14235

## Problem Summary(Required) ：

fix some inconsistent behavior of JDBC external table with version 2.4, which is introduced by https://github.com/StarRocks/starrocks/pull/9944

1. when the length of string type exceeds the limit, the query will return an error in version 2.4, but the previous optimization doesn't handle this case, so I add checking logic to do this after we get_result_from_boxed_array.

2. in some cases, the java class type of result in ResultSetMetadata is not consistent with it in real return data.
for example,FLOAT type in oracle gives java.lang.Double in resultSetMetaData, but the result type is BigDecimal when we getObject from resultSet.
in version 2.4, we get the value by calling `doubleValue()` method through JNI, avoiding this problem.
But after the previous optimization, JDBCScanner will create an object array according to the type in ResultSetMetadata, and an error will occur when assigning BigDecimal to Double.
I made some changes to getNextChunk to fix it by casting the result object to the target type as before.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by silverbullet233** [comment]
run starrocks_admit_test

**Review by kangkaisen — DISMISSED** [review]

**Comment by silverbullet233** [comment]
run starrocks_be_unittest

**Comment by silverbullet233** [comment]
run starrocks_admit_test

File: be/src/exec/vectorized/jdbc_scanner.cpp [hunk 1]
```diff
@@ -390,10 +391,27 @@ Status JDBCScanner::_fill_chunk(jobject jchunk, size_t num_rows, ChunkPtr* chunk
...
             LOCAL_REF_GUARD_ENV(env, jcolumn);
-            auto result_column = _result_chunk->columns()[i].get();
-            auto st = helper.get_result_from_boxed_array(_result_column_types[i], result_column, jcolumn, num_rows);
+            auto& result_column = _result_chunk->columns()[i];
+            auto st =
+                    helper.get_result_from_boxed_array(_result_column_types[i], result_column.get(), jcolumn, num_rows);
             RETURN_IF_ERROR(st);
-            down_cast<NullableColumn*>(result_column)->update_has_null();
+            // check data's length for varchar type
+            if (_slot_descs[i]->type().type == TYPE_VARCHAR) {
```
Comment (stdpain) [inline be/src/exec/vectorized/jdbc_scanner.cpp:399]: we already have a similar function StringColumnWriter::check_string_lengths. we should try to use the same function

Comment (silverbullet233) [reply be/src/exec/vectorized/jdbc_scanner.cpp:399]: > we already have a similar function StringColumnWriter::check_string_lengths. we should try to use the same function

This function itself is not exposed to the outside. although these two functions are similar, but the required params and the returned error message content are different. There is no need to reuse it for reuse, right?

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**